ykcompile/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+                    reg
+                }
+            }
+        }
+    }
+
+    pub fn local_usize(&mut self, a: u32, c: usize) {
+        let reg = self.local_to_reg(a);
+        dynasm!(self.asm
+            ; mov Rq(reg), c as i32
+        );
+    }
+
+    pub fn local_u8(&mut self, a: u32, c: u8) {
+        let reg = self.local_to_reg(a);
+        dynasm!(self.asm
+            ; mov Rq(reg), c as i32
+        );
+    }
+
+    fn local_local(&mut self, var1: u32, var2: u32) {
+        let lreg = self.local_to_reg(var1);
+        let rreg = self.local_to_reg(var2);
+        dynasm!(self.asm
+            ; mov Rq(lreg), Rq(rreg)
+        );
+    }
+
+    fn statement(&mut self, stmt: &Statement) {
+        match stmt {
+            Statement::Assign(l, r) => {
+                let local = l.local.0;
+                match r {
+                    Rvalue::Use(Operand::Place(p)) => {
+                        self.local_local(local, p.local.0)
+                    },
+                    Rvalue::Use(Operand::Constant(c)) => {
+                        match c {
+                            Constant::Int(ci) => {
+                                match ci {
+                                    ConstantInt::UnsignedInt(UnsignedInt::U8(i)) => {
+                                        self.local_u8(local, *i)
+                                    },
+                                    ConstantInt::UnsignedInt(UnsignedInt::Usize(i)) => {
+                                        self.local_usize(local, *i)
+                                    },
+                                    e => panic!("SignedInt, etc: {}", e)
+                                }
+                            }
+                            Constant::Bool(b) => {
+                                self.local_u8(local, *b as u8)
+                            }
+                            c => panic!("Not implemented: {}", c)
+                        }
+                    }
+                    unimpl => panic!("Not implemented: {:?}", unimpl)
+                };
+            },
+            Statement::Return => {},
+            Statement::Nop => {},
+            Statement::Unimplemented(mir_stmt) => println!("Can't compile: {}", mir_stmt)
+        }
+    }
+
+    fn finish(mut self) -> dynasmrt::ExecutableBuffer {
+        dynasm!(self.asm
+            ; ret
+        );
+        self.asm.finalize().unwrap()
+    }
+
+    pub fn compile(tt: TirTrace) -> CompiledTrace {
+        let regs = vec![11, 10, 9, 8, 2, 1];
+        let assembler = dynasmrt::x64::Assembler::new().unwrap();
+        let mut tc = TraceCompiler {
+            asm: assembler,
+            available_regs: regs,
+            assigned_regs: HashMap::new()
+        };
+        for i in 0..tt.len()-1 {
+            let t = tt.op(i);
+            match t {
+                TirOp::Statement(st) => tc.statement(st),
+                TirOp::Guard(_) => { println!("Not implemented: Guard") }
```

Comment:
Don't let Laurie see that `println!`!

Reply:
We used to just print them so we could still compile stuff without implementing guards. Though I'm not sure if guards still appear in our traces.

Reply:
We could emit an instruction which aborts execution, or a nop?

The former stops the trace, the latter allows it to continue.

Reply:
I don't think we want to abort since it's gonna be a while before we implement guards.

Reply:
OK, so let's add a NOP for now. Then at least we can (potentially) see an unimplemented instruction in the code.

Reply:
Like this? b515975

Reply:
Something like that, yes.

Currently we crash on unimplemented statements. Is that what we want?

Reply:
Edd knows me well. But now I've seen it, it must go!

Reply:
It's gone already. As for unimplemented statements, I think for now we should crash. We can change it later if that means we'll be stuck. I will try and keep the tests small enough that they'll allow us to implement one statement at a time.

Reply:
Works for me!

[For some reason when I looked at the code I didn't see all these extra comments. Ah, GitHub updates :/]